src/engine/waypoint_system.cpp

```diff
@@ -0,0 +1,105 @@
+#include "platform/i_platform.h"
+#include "waypoint_system.h"
+#include "core/i_waypoint_component.h"
+#include "core/i_position_component.h"
+#include "core/i_collision_component.h"
+#include "collision_system.h"
+#include "engine.h"
+#include "core/i_controller_component.h"
+#include "core/player_controller_component.h"
+#include "ui/ui.h"
+#include "core/i_health_component.h"
+
+namespace engine {
+
+WaypointSystem::WaypointSystem()
+    : mScene( Scene::Get() )
+{
+}
+
+
+void WaypointSystem::Init()
+{
+    mScene.AddValidator( GetType_static(), []( Actor const& actor )->bool {
+        return actor.Get<IWaypointComponent>().IsValid(); } );
+    mOnMapStart = EventServer<core::MapStartEvent>::Get().Subscribe( boost::bind( &WaypointSystem::OnMapStart, this, _1 ) );
+}
+
+
+void WaypointSystem::Update(double DeltaTime)
+{
+    if (mProgramState.mMode != core::ProgramState::Client)
+    {
+        for (auto actor : mScene.GetActorsFromMap( GetType_static() ))
+        {
+            auto waypointC = actor->Get<IWaypointComponent>();
+            auto positionC( actor->Get<IPositionComponent>() );
+            if (!positionC.IsValid())
+            {
+                return;
+            }
+            auto collisionC( actor->Get<ICollisionComponent>() );
+            if (!collisionC.IsValid())
+            {
+                return;
+            }
+            static auto collisionSystem( ::engine::Engine::Get().GetSystem<engine::CollisionSystem>() );
+            auto&& players( collisionSystem->GetAllCollidingActors(
+                glm::vec2( positionC->GetX(), positionC->GetY() ),
+                collisionC->GetRadius(),
+                1 << CollisionClass::Player ) );
+            auto cnt = std::count_if( mProgramState.mClientDatas.begin(), mProgramState.mClientDatas.end(),
+                [&]( core::ClientData const& clientData )
+            {
+                auto player( mScene.GetActor( clientData.mClientActorGUID ) );
+                if (!player.IsValid())
+                {
+                    return false;
+                }
+                auto healthC( player->Get<IHealthComponent>() );
+                if (!healthC.IsValid())
+                {
+                    return false;
+                }
+
+                return healthC->IsAlive();
+            } );
+            if (players.size() >= cnt) //there could be dead players on the waypoint too
```

Comment:
// TODO this is pretty effin ugly
though i guess not final?
I'd prefer to have the player check/lookup the activatable object, and handle it that way, and not the objects polling nearby player keypresses.
Upside: you can have multiple activatable objects / types in player activation range ( pick up The Lucky Skull of Infinite Luck, step on the portal, push the Unpushable Button of Imminent Doom That Says Don't Even Think About Pushing This Unpushable Button ), mouseover / proximity / whatevere could highlight the selected object, then E or w/e activates it.
Don't see any downsides with that.